### DISCUSSION TITLE
[Panache] Support for keyset pagination?

### DISCUSSION THREAD

**Discussion by agreedSkiing** [quarkusio/quarkus#33606] [Q&A]
Any plans to support a easy to use keyset pagination implementation?

https://vladmihalcea.com/sql-seek-keyset-pagination/

[a comment by an automated account is not shown]

**Comment by FroMage** [comment]
I'm not sure I understood what the page in question describes. Perhaps @Sanne or @gavinking have an opinion?

**Reply by FroMage** [reply]
I'm not sure why this is closed, but I'd still like to hear @Sanne or @gavinking's opinions.

**Reply by Sanne** [reply]
It's a nice feature, which indeed feels like could fit in the spirit of Panache.

I might be wrong but I don't think we have an explicit API for it in Hibernate - you should be able to offer one on top though? I know Blaze does this.

cc/ @beikov

**Reply by beikov** [reply]
Since ORM 6.5 we have `SelectionQuery#getKeyedResultList` that can be used by Panache. Not sure what the best way would be to offer this within Panache, but the basic mechanism is ready to be used :)

**Reply by margaretjoanmiller** [reply]
Sorry if this is a noob question but how does that work? I can't find the documentation of that method, at least none that's clear to me

**Reply by beikov** [reply]
You create a `KeyedPage` and retrieve the keyed results:

```java
KeyedPage<Person> page = Page.first(5).keyedBy(Order.asc(Person.class, "firstName");
KeyedResultList<Person> list = session.createSelectionQuery("from Person", Person.class).getKeyedResultList(page);
```

From there, you can work with the `KeyedPage` that is provided by the `KeyedResultList`:

```java
KeyedPage<Person> page = list.nextPage();
KeyedResultList<Person> nextList = session.createSelectionQuery("from Person", Person.class).getKeyedResultList(page);
```

**Comment by agreedSkiing** [comment]
Well been reading more about this and noticed more and more that it will be up to the develop to design the table to be able to use keyset pagination and the develop to implement the query

**Comment by margaretjoanmiller** [comment]
I would also love to be able to do this natively in panache!
